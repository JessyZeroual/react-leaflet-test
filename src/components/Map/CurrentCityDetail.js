import React from "react";
import {
  StyledWrapperCurrentCityDetail,
  StyledCurrentCityDetailText,
} from "./Map.styled";

const CurrentCityDetail = ({ currentCity }) => (
  <>
    <StyledWrapperCurrentCityDetail>
      <img src="/icon/buildings.svg" width="30" alt="icon city" />
      <StyledCurrentCityDetailText>
        {currentCity.city}
      </StyledCurrentCityDetailText>
    </StyledWrapperCurrentCityDetail>
    <StyledWrapperCurrentCityDetail>
      <img src="/icon/flag.svg" width="30" alt="icon state" />
      <StyledCurrentCityDetailText>
        {currentCity.state}
      </StyledCurrentCityDetailText>
    </StyledWrapperCurrentCityDetail>
    <StyledWrapperCurrentCityDetail>
      <img src="/icon/growth.svg" width="30" alt="icon growth" />

      <StyledCurrentCityDetailText
        isNegative={Math.sign(currentCity.growth_from_2000_to_2013) === -1}
      >
        {`${currentCity.growth_from_2000_to_2013}%`}
      </StyledCurrentCityDetailText>
    </StyledWrapperCurrentCityDetail>
    <StyledWrapperCurrentCityDetail>
      <img src="/icon/population.svg" width="30" alt="icon population" />
      <StyledCurrentCityDetailText>
        {currentCity.population.toLocaleString("en-US")}
      </StyledCurrentCityDetailText>
    </StyledWrapperCurrentCityDetail>
  </>
);

export default CurrentCityDetail;
