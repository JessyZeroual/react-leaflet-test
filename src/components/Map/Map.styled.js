import styled from "styled-components";

export const StyledWrapperMap = styled.div`
  border: 1px solid grey;
  margin-left: 20px;
  flex: 1;
`;

export const StyledWrapperCurrentCityDetail = styled.div`
  display: flex;
  align-items: center;
`;

export const StyledCurrentCityDetailText = styled.p`
  font-weight: bold;
  font-size: 15px;
  padding: 0px 10px;
  color: ${({ isNegative }) => isNegative && "red"};
`;
