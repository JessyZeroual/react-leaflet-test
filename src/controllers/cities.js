export const getCities = () => {
  return fetch(`/cities.json`)
    .then((res) => res.json())
    .then((data) =>
      data.map((city) => ({
        ...city,
        growth_from_2000_to_2013: Number(
          city.growth_from_2000_to_2013.replace("%", "")
        ),
        population: Number(city.population),
        rank: Number(city.rank),
      }))
    );
};
