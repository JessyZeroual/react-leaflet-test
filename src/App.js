import React from "react";

import NavigationBar from "./components/NavigationBar/NavigationBar";
import Map from "./components/Map/Map";
import useCities from "./helpers/useCities";
import { CitiesContextProvider } from "./context/CitiesContext";
import { StyledApp } from "./App.styled";

const App = () => {
  const { isLoading, cities, currentCity, setCurrentCity } = useCities();

  return (
    <CitiesContextProvider
      value={{
        state: { isLoading, cities, currentCity },
        dispatch: { setCurrentCity },
      }}
    >
      <StyledApp>
        <NavigationBar />
        <Map currentCity={currentCity} />
      </StyledApp>
    </CitiesContextProvider>
  );
};

export default App;
