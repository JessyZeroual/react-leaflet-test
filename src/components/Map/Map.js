import React, { useState, useEffect } from "react";
import { Map as LeafletMap, Marker, Popup, TileLayer } from "react-leaflet";
import { StyledWrapperMap } from "./Map.styled";
import CurrentCityDetail from "./CurrentCityDetail";

const Map = ({ currentCity }) => {
  const [zoom, setZoom] = useState(7);
  const [latitude, setLatitude] = useState(40.7127837);
  const [longitude, setLongitude] = useState(-74.0059413);

  useEffect(() => {
    const animation = () => {
      setLatitude(currentCity.latitude);
      setLongitude(currentCity.longitude);
      setTimeout(() => {
        setZoom(7);
      }, 800);
    };
    setZoom(4);
    if (currentCity.latitude) animation();
  }, [currentCity]);

  return (
    <StyledWrapperMap>
      <LeafletMap center={[latitude, longitude]} zoom={zoom} minZoom={2}>
        <TileLayer url="http://{s}.tile.osm.org/{z}/{x}/{y}.png" />
        <Marker position={[latitude, longitude]}>
          <Popup closeButton={false}>
            <CurrentCityDetail currentCity={currentCity} />
          </Popup>
        </Marker>
      </LeafletMap>
    </StyledWrapperMap>
  );
};

export default Map;
