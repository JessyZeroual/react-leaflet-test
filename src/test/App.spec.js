import React from "react";
import TestRenderer from "react-test-renderer";
import { act } from "react-test-renderer";
import fetchMock from "fetch-mock";

import App from "../App";
import Map from "../components/Map/Map";
import cities from "./listOfCitiesTests";

jest.mock("leaflet");
const elementMock = {
  addEventListener: jest.fn(),
  removeEventListener: jest.fn(),
};
jest.spyOn(document, "getElementById").mockImplementation(() => elementMock);

describe("App", () => {
  let renderer;
  beforeEach(async () => {
    await act(async () => {
      renderer = TestRenderer.create(<App />);
      fetchMock.get("/cities.json", cities, { overwriteRoutes: true });
    });
  });

  describe("When App is mounted", () => {
    it("renders correctly", () => {
      expect(renderer).toMatchSnapshot();
    });

    it("Pass the props currentCity to the component Map", () => {
      expect(renderer.root.findByType(Map).props.currentCity).toEqual({
        city: "New York",
        growth_from_2000_to_2013: 4.8,
        latitude: 40.7127837,
        longitude: -74.0059413,
        population: 8405837,
        rank: 1,
        state: "New York",
      });
    });
  });

  describe("When we click on a city", () => {
    it("Pass the new props currentCity to the component Map", async () => {
      await act(async () => {
        const button = renderer.toJSON().children[0].children[1].children[1];
        button.props.onClick();
      });

      expect(renderer.root.findByType(Map).props.currentCity).toEqual({
        city: "Los Angeles",
        growth_from_2000_to_2013: 4.8,
        latitude: 34.0522342,
        longitude: -118.2436849,
        population: 3884307,
        rank: 2,
        state: "California",
      });
    });
  });
});
