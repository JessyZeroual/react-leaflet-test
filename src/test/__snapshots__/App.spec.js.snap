// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App When App is mounted renders correctly 1`] = `
<div
  className="sc-fznZeY jiuyPv"
>
  <aside
    className="sc-AxmLO leUibO"
  >
    <div
      className="sc-fzozJi iDouY"
    >
      <form
        className="sc-AxgMl hyRgag"
        onSubmit={[Function]}
      >
        <div
          className="sc-AxhUy iWNaag"
        >
          <img
            alt="icon search"
            src="/icon/search.svg"
            width="15"
          />
          <input
            className="sc-AxhCb itDVzo"
            onChange={[Function]}
          />
        </div>
        
      </form>
    </div>
    <div
      className="sc-fzoLsD eBOzkn"
      id="mainSideBarID"
    >
      <div
        className="sc-AxjAm eQRonO"
        onClick={[Function]}
      >
        New York
      </div>
      <div
        className="sc-AxjAm bbjfxF"
        onClick={[Function]}
      >
        Los Angeles
      </div>
      <div
        className="sc-AxjAm bbjfxF"
        onClick={[Function]}
      >
        Chicago
      </div>
      <div
        className="sc-AxjAm bbjfxF"
        onClick={[Function]}
      >
        Houston
      </div>
      <div
        className="sc-AxjAm bbjfxF"
        onClick={[Function]}
      >
        Philadelphia
      </div>
      <div
        className="sc-AxjAm bbjfxF"
        onClick={[Function]}
      >
        Phoenix
      </div>
      <div
        className="sc-AxjAm bbjfxF"
        onClick={[Function]}
      >
        San Antonio
      </div>
      <div
        className="sc-AxjAm bbjfxF"
        onClick={[Function]}
      >
        San Diego
      </div>
      <div
        className="sc-AxjAm bbjfxF"
        onClick={[Function]}
      >
        Dallas
      </div>
      <div
        className="sc-AxjAm bbjfxF"
        onClick={[Function]}
      >
        San Jose
      </div>
      <div
        className="sc-AxjAm bbjfxF"
        onClick={[Function]}
      >
        Austin
      </div>
      <div
        className="sc-AxjAm bbjfxF"
        onClick={[Function]}
      >
        Indianapolis
      </div>
      <div
        className="sc-AxjAm bbjfxF"
        onClick={[Function]}
      >
        Jacksonville
      </div>
      <div
        className="sc-AxjAm bbjfxF"
        onClick={[Function]}
      >
        San Francisco
      </div>
      <div
        className="sc-AxjAm bbjfxF"
        onClick={[Function]}
      >
        Columbus
      </div>
      <div
        className="sc-AxjAm bbjfxF"
        onClick={[Function]}
      >
        Charlotte
      </div>
      <div
        className="sc-AxjAm bbjfxF"
        onClick={[Function]}
      >
        Fort Worth
      </div>
      <div
        className="sc-AxjAm bbjfxF"
        onClick={[Function]}
      >
        Detroit
      </div>
      <div
        className="sc-AxjAm bbjfxF"
        onClick={[Function]}
      >
        El Paso
      </div>
      <div
        className="sc-AxjAm bbjfxF"
        onClick={[Function]}
      >
        Memphis
      </div>
    </div>
  </aside>
  <div
    className="sc-fzplWN sDynr"
  >
    <div />
  </div>
</div>
`;
